upup/pkg/kutil/rollingupdate_cluster.go

```diff
@@ -269,35 +281,169 @@ func buildCloudInstanceGroup(ig *api.InstanceGroup, g *autoscaling.Group, nodeMa
 	return n
 }
 
-func (n *CloudInstanceGroup) RollingUpdate(cloud fi.Cloud, force bool, interval time.Duration, k8sClient *k8s_clientset.Clientset) error {
-	c := cloud.(awsup.AWSCloud)
+// TODO: Temporarily increase size of ASG?
+// TODO: Remove from ASG first so status is immediately updated?
+// TODO: Batch termination, like a rolling-update
+
+// RollingUpdateDrainValidate performs a rolling update on a list of ec2 instances.
+func (n *CloudInstanceGroup) RollingUpdate(rollingUpdateData *RollingUpdateCluster, instanceGroupList *api.InstanceGroupList, isBastion bool, t time.Duration) (err error) {
+
+	// we should not get here, but hey I am going to check
+	if rollingUpdateData == nil {
+		return fmt.Errorf("rollingUpdate cannot be nil")
+	}
+
+	// Do not need a k8s client if you are doing cloud only
+	if rollingUpdateData.K8sClient == nil && !rollingUpdateData.CloudOnly {
+		return fmt.Errorf("rollingUpdate is missing a k8s client")
+	}
+
+	if instanceGroupList == nil {
+		return fmt.Errorf("rollingUpdate is missing the InstanceGroupList")
+	}
+
+	c := rollingUpdateData.Cloud.(awsup.AWSCloud)
 
 	update := n.NeedUpdate
-	if force {
+	if rollingUpdateData.Force {
 		update = append(update, n.Ready...)
 	}
+
+	if isBastion || rollingUpdateData.CloudOnly {
+		glog.Infof("Not validating the cluster")
+	} else if rollingUpdateData.FailOnValidate && featureflag.DrainAndValidateRollingUpdate.Enabled() {
+		if err = n.ValidateCluster(rollingUpdateData, instanceGroupList); err != nil {
+			glog.Errorf("Error validating cluster: %s.", err)
+			return err
+		}
+	}
+
 	for _, u := range update {
-		instanceID := aws.StringValue(u.ASGInstance.InstanceId)
-		glog.Infof("Stopping instance %q in AWS ASG %q", instanceID, n.ASGName)
 
-		// TODO: Evacuate through k8s first?
+		if isBastion {
+
+			if err = n.DeleteAWSInstance(u, c); err != nil {
+				return err
+			}
+			glog.Infof("Updated a bastion instance and continuing.")
+
+			continue
+		}
+
+		if rollingUpdateData.CloudOnly {
+			glog.Warningf("Not draining nodes - cloud only is set.")
+		} else if featureflag.DrainAndValidateRollingUpdate.Enabled() {
+
+			glog.Infof("Draining the node.")
+
+			if err = n.DrainNode(u, rollingUpdateData); err == nil {
+				glog.Errorf("Error draining node: %s.", err)
```

Comment:
You should just `return fmt.Errorf("error draining node %q: %v", nodeName, err)`